File: common/src/main/java/net/impactdev/gts/common/discord/internal/DiscordSourceSpecificPlaceholderParser.java
package net.impactdev.gts.common.discord.internal;

import net.impactdev.impactor.api.placeholders.PlaceholderSources;

import java.util.List;
import java.util.Objects;
import java.util.function.Function;
import java.util.function.Supplier;

public class DiscordSourceSpecificPlaceholderParser<T> implements DiscordPlaceholderParser {

    private final Class<T> sourceType;
    private final String id;
    private final Function<T, String> parser;

    public DiscordSourceSpecificPlaceholderParser(Class<T> sourceType, String id, Function<T, String> parser) {
        this.sourceType = sourceType;
        this.id = id;
        this.parser = parser;
    }

    @Override
    public String getID() {
        return this.id;
    }

    @Override
    public String parse(PlaceholderSources sources) {
        return sources.suppliers().stream()
                .map(Supplier::get)
                .filter(x -> this.sourceType.isAssignableFrom(x.getClass()))
                .map(this.sourceType::cast)
                .map(this.parser)
                .filter(Objects::nonNull)
                .findAny()
                .orElse(null);
    }

}

File: common/src/main/kotlin/net/impactdev/gts/common/discord/internal/DiscordSourceSpecificPlaceholderParser.kt
package net.impactdev.gts.common.discord.internal

import java.util.*
import java.util.function.Function
import java.util.function.Supplier

class DiscordSourceSpecificPlaceholderParser<T>(
    private val sourceType: Class<T>,
    override val iD: String,
    private val parser: Function<T, String?>
) : DiscordPlaceholderParser {
    override fun parse(sources: List<Supplier<Any>>): String? {
        return sources.stream()
            .map { obj: Supplier<Any> -> obj.get() }
            .filter { x: Any -> sourceType.isAssignableFrom(x.javaClass) }
            .map { obj: Any? -> sourceType.cast(obj) }
            .map(parser)
            .filter { obj: String? -> Objects.nonNull(obj) }
            .findAny()
            .orElse(null)
    }
}